nexus/networking/src/firewall_rules.rs

```diff
@@ -439,3 +486,100 @@ pub async fn plumb_service_firewall_rules(
+///
+/// This looks up the source IP allowlist in the database, and generates
+/// firewall rules attached to each provided NIC that implements the allowlist.
+/// If the allowlist is "any", there are no restrictions and no rules are
+/// returned.
+///
+/// Otherwise, two sets of rules are returned:
+///
+/// - A low-priority rule that denies all traffic
+/// - A higher-priority rule that allows it from the specific IPs.
+async fn allowlist_to_firewall_rules(
+    nics: &[NetworkInterface],
+    allowed_ips: Vec<IpNet>,
+) -> Option<Vec<sled_agent_client::types::VpcFirewallRule>> {
+    if nics.is_empty() {
+        return None;
+    }
+
+    // Convert to host identifiers, how the sled-agent types expose these.
+    assert!(!allowed_ips.is_empty(), "Allowlist must not be empty here");
+    let filter_hosts = Some(
+        allowed_ips
+            .into_iter()
+            .map(|net| sled_agent_client::types::HostIdentifier::Ip(net.into()))
+            .collect(),
+    );
+
+    let low_priority_deny = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts: filter_hosts.clone(),
+        filter_ports: Some(vec![L4PortRange {
+            first: 1.try_into().unwrap(),
+            last: u16::MAX.try_into().unwrap(),
+        }
+        .into()]),
+        filter_protocols: None,
+        action: VpcFirewallRuleAction::Deny.into(),
+        priority: ALLOW_LIST_LOW_PRIORTY,
+    };
+
+    let allow = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts,
+
+        // TODO-correctness: Having no filter on ports or protocols is not
+        // techniically correct. Nexus should only receive TCP traffic, on HTTP
+        // or HTTPs traffic, for example. However, at this point in the system,
+        // we don't really have access to _which service_ this NIC is for, only
+        // that it is for _one_ of our services.
+        //
+        // We may want to reorganize this code to apply these rules on a
+        // per-service basis. This is discussed in RFD 474.
+        filter_ports: None,
+        filter_protocols: None,
+        action: VpcFirewallRuleAction::Allow.into(),
+        priority: ALLOWLIST_HIGH_PRIORITY,
+    };
+
+    Some(vec![low_priority_deny, allow])
+}
+
+/// Return the list of allowed IPs from the database.
+async fn lookup_allowed_source_ips(
+    datastore: &DataStore,
+    opctx: &OpContext,
+    log: &Logger,
+) -> Result<AllowedSourceIps, Error> {
+    match datastore.allowed_source_ips_view(opctx).await {
+        Ok(allowed) => {
+            debug!(log, "fetched allowlist from DB"; "allowed" => ?allowed);
+            let ips = if let Some(list) = allowed.allowed_ips {
```

Comment:
Same comment about using `match` over `if/let` here. 

Actually, now that I write that, do we have a conversion from the db type of `allowed` here into `AllowedSourceIps` already that we could call instead?

Reply:
Yes, we have a conversion like that (or I can write one easily). Will change!

Reply:
Added in 381b1d2c0